listener/pipeline.go

```diff
@@ -0,0 +1,142 @@
+package listener
+
+import (
+	"runtime"
+	"sync"
+)
+
+type pipe struct {
+	in            jobChan
+	out           jobChan
+	processErrors bool
+
+	queue     chan queuedJobChan
+	closeInCh chan struct{}
+}
+
+// Job is a short callback signature, used in pipes
+type Job func(msg interface{}) interface{}
+
+type (
+	jobChan       chan interface{}
+	queuedJobChan chan interface{}
+)
+
+func newPipe(job Job, concurrency int, processErrors bool) *pipe {
+	p := &pipe{
+		in:            make(jobChan, 1),
```

Comment:
Why are all of the channels buffered of size 1?

They will block as long as there is a message already in the queue